File: packages/webview_lib/lib/src/rpc/rpc_call.dart
import 'dart:async';

import 'interface.dart';

late IMediaPlatform platform;

Future<List<dynamic>> rpc(List<IRpcFnc> calls) {
  final msg = IRpc(rpcId: lastPromiseIdx++, fncs: calls);
  print('flutter rpc (rpcId=${msg.rpcId})');
  final comp = promises[msg.rpcId] = Completer<List<dynamic>>();
  platform.postToWebView(msg);
  return comp.future;
}

void _rpcCallback(IRpcResult rpcResult) {
  print('flutter rpc Callback (rpcId=${rpcResult.rpcId})');
  final comp = promises.remove(rpcResult.rpcId);
  if (comp == null) throw Exception('not found');
  if (rpcResult.error != null)
    comp.completeError(rpcResult.error!);
  else
    comp.complete(rpcResult.result);
}

void handlerCallback(IStreamMessage msg) {
  if (msg.handlerId == null) return;
  final listenner = handlerListenners[msg.handlerId];
  if (listenner == null) return;
  listenner(msg.streamId, msg.value);
}

void receiveFromWebView(IStreamMessage msg) {
  switch (msg.streamId) {
    case StreamIds.consoleLog:
      print(msg.value);
      break;
    case StreamIds.rpcCallback:
      _rpcCallback(IRpcResult.fromJson(msg.value as Map<String, dynamic>));
      break;
    default:
      handlerCallback(msg);
      break;
  }
}

final handlerListenners = <int, void Function(int streamId, dynamic value)>{};

final promises = <int, Completer>{};
var lastPromiseIdx = 1;

File: browser/typescript/lib/src/htmlApp/rpc_call.ts
import { IRpcFnc, IStreamMessage, IRpcResult, IRpc, receivedFromFlutter, StreamIds } from "../rpc/index";

export function rpc(calls: IRpcFnc[]): Promise<any[]> {
    let msg: IRpc = { rpcId: lastPromiseIdx++, fncs: calls };
    console.log(`html rpc (rpcId=${msg.rpcId})`)
    return new Promise<any[]>((resolve, reject) => {
        promises[msg.rpcId.toString()] = { resolve: resolve, reject: reject };
        callJavascript(`wikib.receivedFromFlutter (${JSON.stringify(msg).replace('\\','\\\\').replace("'", "\'")})`);
    });
}

export function callJavascript(script: string): Promise<void> {
    eval(script)
    return Promise.resolve()
}

let promises: { [idx: string]: IResolveReject | undefined } = {};
let lastPromiseIdx = 1;

export function receiveFromWebView(msg: IStreamMessage<any>) {
    switch (msg.streamId) {
        case StreamIds.rpcCallback:
            rpcCallback(msg)
            break
        case StreamIds.consoleLog:
            break
        default:
            handlerCallback(msg)
            break
    }
}
export let handlerListenners: { [name: string]: (streamId: StreamIds, value: IRpcResult<any>) => void } = {}

function handlerCallback(msg: IStreamMessage<IRpcResult<any>>) {
    if (!msg.handlerId) return
    let listenner = handlerListenners[msg.handlerId]
    if (!listenner) return
    listenner(msg.streamId, msg.value)
}

function rpcCallback(msg: IStreamMessage<IRpcResult<any>>) {
    console.log(`flutter rpc Callback (rpcId=${msg.value.rpcId})`)
    let resolveReject = promises[msg.value.rpcId.toString()]
    delete promises[msg.value.rpcId.toString()]
    if (!resolveReject) throw 'not found'
    if (msg.value.error != null) resolveReject.reject(msg.value.error);
    else resolveReject.resolve(msg.value.result);
}

interface IResolveReject {
    resolve: (res: any) => void
    reject: (error?: any) => void
}

